piccolo_api/jwt_auth/middleware.py

```diff
@@ -129,12 +136,13 @@ async def __call__(self, scope, receive, send):
         is recognised, otherwise raise a 403 HTTP error.
         """
         allow_unauthenticated = self.allow_unauthenticated
+        visible_paths = self.visible_paths
 
         headers = dict(scope["headers"])
         token = self.get_token(headers)
         if not token:
             error = JWTError.token_not_found.value
-            if allow_unauthenticated:
+            if allow_unauthenticated or visible_paths:
```

Comment:
You need to be careful here because this is checking to the presence of `visible_paths`, and not whether it matches the current path.

Reply:
Do you think `scope["path"] in visible_paths` is a better option?

Reply:
I think checking presence of `visible_paths` (if is empty list is or not specified we got `Token not found` error) enough because we can make a GET request only for Swagger docs from browser on routes `/docs` and `/openapi.json` where we have to authorize the user if we want to communicate with the endpoints. All other routes must have a token in the header.

Reply:
> Do you think scope["path"] in visible_paths is a better option?

Yes, something like that.

Reply:
OK, I'll change that even though I think it might not be necessary (see previous comment) because if for example we pass a `/bands` route to `visible_paths`, we can't make a request to that route directly from the browser (we'll get `Token not found`). Basically we can only pass the Swagger docs routes `/docs` and `/openapi.json`. Sorry if I'm wrong.

Reply:
If `visible_paths` matches the current path, then we should just let it bypass all other checks. I think this is the problem - it's still checking for a token etc, but we don't need to.

Reply:
Put the check for `visible_paths` at the top level.

Reply:
```python
    async def __call__(self, scope, receive, send):
        """
        Add the user_id to the scope if a JWT token is available, and the user
        is recognised, otherwise raise a 403 HTTP error.
        """
        allow_unauthenticated = self.allow_unauthenticated
        visible_paths = self.visible_paths
        
        if scope['path'] in self.visible_paths:
            await self.asgi(scope, receive, send)
        
        # then all of the existing code
```

Reply:
I'd maybe change `visible_paths` to something like `excluded_paths`. Meaning it's excluded from the checks.

Reply:
Yeah. All the time I tested with
```python
auth_header = APIKeyHeader(name="Authorization")
private_app = FastAPI(dependencies=[Depends(auth_header)])
```
and with `auth_header` as dependency to the whole `private_app` only Swagger endpoints work in `visible_paths` because the authorization header dependency expected header on every request (that's why the `/bands` endpoint requested a token). I'm sorry. 
I will make the changes you suggested. If that's not enough, feel free to change anything.

Reply:
Ok, interesting. I didn't consider the interaction with `APIKeyHeader`.

It might be best to attach `APIKeyHeader` to an `APIRouter` rather than the app.

Reply:
Yes. We can do this at `APIRouter` level or if we use `FastAPIWrapper` we can do this
```python
auth_header = APIKeyHeader(name="Authorization")
private_app = FastAPI()

FastAPIWrapper(
    "/bands/",
    fastapi_app=private_app,
    piccolo_crud=PiccoloCRUD(Band, read_only=False),
    fastapi_kwargs=FastAPIKwargs(
        all_routes={"tags": ["Band"]},
        post={"dependencies": [Depends(auth_header)]},  # protected route
        put={"dependencies": [Depends(auth_header)]},  # protected route
        patch={"dependencies": [Depends(auth_header)]},  # protected route
        delete_single={
            "dependencies": [Depends(auth_header)]
        },  # protected route
    ),
)
```
But to be honest, I like it best at the application level, as we do in Piccolo Admin, because a protected (private) application should be private, and I think the best option is that only Swagger docs can be seen (for frontend developers as requested in issue) but in order to communicate with the endpoints, the user must be authorized which is provided with `APIKeyHeader`.